src/main/java/xyz/nkomarn/harbor/listener/AfkListener.java

```diff
@@ -0,0 +1,153 @@
+
+        // Register listeners after populating the queue
+        Bukkit.getServer().getPluginManager().registerEvents(this, plugin);
+
+        // We want every player to get a check every 20 ticks.
+        movementChecker.runTaskTimer(plugin, 20, 20);
+
+        JavaPlugin.getPlugin(Harbor.class).getLogger().info("Fallback AFK detection system is enabled");
+    }
+
+    /**
+     * Provides a way to halt the listener
+     */
+    public void stop() {
+        movementChecker.cancel();
+        HandlerList.unregisterAll(this);
+        playerQueue = null;
+        JavaPlugin.getPlugin(Harbor.class).getLogger().info("Fallback AFK detection system is disabled");
+    }
+
+    @EventHandler(ignoreCancelled = true)
+    public void onChat(AsyncPlayerChatEvent event) {
+        afkProvider.updateActivity(event.getPlayer());
+    }
+
+    @EventHandler(ignoreCancelled = true)
+    public void onCommand(PlayerCommandPreprocessEvent event) {
+        afkProvider.updateActivity(event.getPlayer());
+    }
+
+    @EventHandler(ignoreCancelled = true)
+    public void onInventoryClick(InventoryClickEvent event) {
+        afkProvider.updateActivity((Player) event.getWhoClicked());
+    }
+
+    @EventHandler
+    public void onJoin(PlayerJoinEvent event) {
+        playerQueue.add(new AfkPlayer(event.getPlayer()));
+        afkProvider.updateActivity(event.getPlayer());
+    }
+
+    @EventHandler
+    public void onLeave(PlayerQuitEvent event) {
+        playerQueue.remove(new AfkPlayer(event.getPlayer()));
+        afkProvider.removePlayer(event.getPlayer().getUniqueId());
+    }
+
+    /**
+     * Internal class for handling the task of checking player movement; Is a separate task so that we can cancel and restart it easily
+     */
+    private class PlayerMovementChecker extends BukkitRunnable {
+        @Override
+        public synchronized void run() {
+            // Short circuit exit if there aren't any players to check
+            if (playerQueue.isEmpty())
+                return;
+
+
+            AfkPlayer firstPlayer = playerQueue.peek();
+            AfkPlayer player = firstPlayer;
+
+            long start = System.currentTimeMillis();
+            do {
+                if (player.changed()) {
+                    afkProvider.updateActivity(player.player);
+                }
+                playerQueue.add(player);
+            } while (System.currentTimeMillis() - start < MAX_PROCESS_TIME_MS &&
+                    !playerQueue.isEmpty() &&
+                    !(player = playerQueue.poll()).equals(firstPlayer));
+        }
+    }
+
+    private static class AfkPlayer {
+        private final Player player;
+        private int hash;
+
+        public AfkPlayer(Player player) {
+            this.player = player;
+        }
+
+        /**
+         * Check if the player changed its position since the last check
+         *
+         * @return true if the position changed
+         */
+        private synchronized boolean changed() {
```

Comment:
please remove synchronized here. this is missleading in a main thread task.

Reply:
Will do!

Reply:
The synchronized keywords were added accidentially due to an incorrect flag set in my IDE, and are now fixed.

I will also do the 20ms cleanup, and, as that condition is unnecessary, we can simply switch to an arraylist

Reply:
![image](https://user-images.githubusercontent.com/4931930/113329500-e22c1e00-92da-11eb-8d9a-23b494c8c147.png)